rbd: write max 1gb per WriteSame() operation

nixpanic opened the pull request:
# Describe what this PR does #

It seems that writing more than 1 GiB per WriteSame() operation causes
an EINVAL (22) "Invalid argument" error. Splitting the writes in blocks
of maximum 1 GiB should prevent that from happening.

Not all volumes are of a size that is the multiple of the stripe-size.
WriteSame() needs to write full blocks of data, so in case there is a
small left-over, it will be filled with WriteAt().

## Testing ##

Manual testing with thick-provisioning was done with different sizes of
volumes, ranging from 50 MB to 250 GB. This uncovered an other issue
when the size of the volume is not a multiple of the stripe-size.

Large thick-provisioned volumes are slow in getting to the Bound state.
Warnings might be logged as Kubernetes Events. These warnings may
not be fatal, provisioning the volume will continue in the background (as
long as the provisioner Pod is not restarted).

Example of the Events:
```
2m31s       Normal    ExternalProvisioning    persistentvolumeclaim/thick-50gb   waiting for a volume to be created, either by external provisioner "rbd.csi.ceph.com" or manually created by system administrator
95s         Normal    Provisioning            persistentvolumeclaim/thick-50gb   External provisioner is provisioning volume for claim "default/thick-50gb"
5m49s       Warning   ProvisioningFailed      persistentvolumeclaim/thick-50gb   failed to provision volume with StorageClass "ceph-rbd-thick": rpc error: code = DeadlineExceeded desc = context deadline exceeded
3m43s       Warning   ProvisioningFailed      persistentvolumeclaim/thick-50gb   failed to provision volume with StorageClass "ceph-rbd-thick": rpc error: code = Aborted desc = an operation with the given Volume ID pvc-ebc02ff0-7e54-4bff-9157-0e7ce4120818 already exists
95s         Normal    ProvisioningSucceeded   persistentvolumeclaim/thick-50gb   Successfully provisioned volume pvc-ebc02ff0-7e54-4bff-9157-0e7ce4120818
```

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
rbd: write max 1gb per WriteSame() operation

It seems that writing more than 1 GiB per WriteSame() operation causes
an EINVAL (22) "Invalid argument" error. Splitting the writes in blocks
of maximum 1 GiB should prevent that from happening.

Not all volumes are of a size that is the multiple of the stripe-size.
WriteSame() needs to write full blocks of data, so in case there is a
small left-over, it will be filled with WriteAt().

Signed-off-by: Niels de Vos <ndevos@redhat.com>

Madhu-1 commented:
@nixpanic is it possible to add E2E for 50 Mib?

humblec commented:
LGTM. however will wait for the e2e comment to be answered before approving..

nixpanic commented:
> @nixpanic is it possible to add E2E for 50 Mib?

Of course, do you want it in this PR or an other one?

Madhu-1 commented:
> > @nixpanic is it possible to add E2E for 50 Mib?
> 
> Of course, do you want it in this PR or an other one?

good to get that in this PR so that functionality will get tested.

nixpanic committed:
e2e: rework thick-provisioning test case

The stripe-size is the most efficient size to write to RBD images.
However, not all images are a multiple of stripe-size large. That means
thick-provisioning would not allocate the full image, and the process
might even fail.

This adds a 50 MB PVC to test the process, 100 MB is coincidentally a
multiple of the (default 4 MB) stripe-size, 50 MB is not.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
job **ci/centos/mini-e2e-helm/k8s-1.19** failed deploying rook with:
```
Error from server: etcdserver: request timed out
```

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19
